Add a pattern for using a shared object as a cache

jaehyeon-kim opened the pull request:
**Please** add a meaningful description for your change here

A cache is a software component that stores data so that future requests for that data can be served faster. Side inputs, stateful `DoFn` and calling an external service can be used for accessing a cache. The Python SDK provide another option in the shared module, which can be more efficient than side inputs in terms of the amount of memory required and simpler than stateful `DoFn`.

The samples on this page show you how to use the Shared class of the [shared module](https://beam.apache.org/releases/pydoc/current/apache_beam.utils.shared.html) to enrich elements in both bounded and unbounded `PCollection`s

It supports this issue - https://github.com/apache/beam/issues/32177

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

liferoad commented on website/www/site/content/en/documentation/patterns/overview.md:
```diff
@@ -54,6 +54,10 @@ Pipeline patterns demonstrate common Beam use cases. Pipeline patterns are based
 * [Grouping elements for efficient external service calls](/documentation/patterns/grouping-elements-for-efficient-external-service-calls/#grouping-elements-for-efficient-external-service-calls-using-the-`GroupIntoBatches`-transform)
 
+**Cache with a shared object** - Patterns for using a shared object as a cache
+* [Create a cache on a batch pipeline](/documentation/patterns/shared-class/#create-a-cache-on-a-batch-pipeline)
```
shall we call it out this only works for Python?

jaehyeon-kim replied:
Let me update to *Patterns for using a shared object as a cache using the Python SDK* in a separate commit.

jaehyeon-kim replied:
Resolved by a separate PR.

damccorm commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+
+# Cache data using a shared object
+
+A cache is a software component that stores data so that future requests for that data can be served faster. Side inputs, stateful `DoFn` and calling an external service can be used for accessing a cache. The Python SDK provide another option in the shared module, which can be more efficient than side inputs in terms of the amount of memory required and simpler than stateful `DoFn`.
```
```suggestion
A cache is a software component that stores data so that future requests for that data can be served faster. Side inputs, stateful `DoFn` and calling an external service can be used for accessing a cache. The Python SDK provide another option in the shared module, which can be more efficient than side inputs in terms of the amount of memory required and simpler than a stateful `DoFn`.
```

Nit

jaehyeon-kim replied:
Resolved by a separate PR.

damccorm commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+
+In this example, the customer cache is loaded as a dictionary in the `setup` method of the `EnrichOrderFn`, and it is used to add customer attributes to the order records. Note that a Shared object encapsulates a weak reference to a singleton instance of the shared resource. Therefore, we need to create a wrapper class since the Python dictionary does not support weak references.
+
+{{< highlight py >}}
```
I think there are a number of pieces we don't need here for our example, and we could greatly simplify by just providing an implementation of `EnrichOrderFn`. So this example would become:

```
class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._version = 1
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers)

    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers(version=self._version)
        return WeakRefDict(self._customers)

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
```

jaehyeon-kim replied:
Let me update with the following snippet in a separate commit.

```python
class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers)

    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers()
        return WeakRefDict(self._customers)

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
```

jaehyeon-kim replied:
Resolved by a separate commit.

damccorm commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+    def setup(self):
+        # setup is a good place to initialize transient in-memory resources.
+        self._customer_lookup = self._shared_handle.acquire(
+            self.load_customers, datetime.now().timestamp()
```
Calling acquire with a timestamp based tag in setup is probably a bad idea since this will be consistently changing for each thread which calls setup.

jaehyeon-kim replied:
Would you consider it is okay if it is changed as

```python
    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers, "init")
```

damccorm replied:
I think using the timestamp bucketized version I mentioned below probably makes more sense (see that thread for explanation of why)

jaehyeon-kim replied:
Resolved by a separate commit.

damccorm commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+    def start_bundle(self):
+        # update the lookup table when the time difference exceeds the threshold.
+        current_ts = datetime.now().timestamp()
+        ts_diff = current_ts - self._customers["timestamp"]
```
Rather than using `ts_diff` which can be different for each thread (causing the cache to be reloaded several times), it would be better to use a more deterministic mapping.

For example, you could do something like:

`tag = current_ts - (current_ts % stale._max_stale_sec)`. This would ensure that you're always still within `max_stale_sec`, but you're not reloading the cache more than needed. There may be an even smarter strategy here using a second shared object (which could tell you, for example, the tag of the latest results), but I think this is probably good enough

jaehyeon-kim replied:
Can you clarify more? `ts_diff` is necessary to check whether the shared object should be refreshed or not. The tag value can be updated to be more deterministic. Would you consider the following snippet is okay?

```python
class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._max_stale_sec = 60
        self._cache_version = 0
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(self.load_customers, self._cache_version)
        self._cache_version +=1
 
    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers()
        return WeakRefDict(self._customers)

    def start_bundle(self):
        # Update the lookup table when the time difference exceeds the threshold.
        current_ts = datetime.now().timestamp()
        ts_diff = current_ts - self._customers["timestamp"]
        if ts_diff > self._max_stale_sec:
            self._customer_lookup = self._shared_handle.acquire(self.load_customers, self._version)
            elf._cache_version +=1

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
```

damccorm replied:
This is probably a bit better, but it still has some potential issues. Imagine the following sequence of events:

```
Worker spins up DoFn1 and DoFn2
DoFn1.setup() acquires shared_handle with _cache_version = 0
DoFn2.setup() acquires shared_handle with _cache_version = 0 (so far, so good)
DoFn1.start_bundle() acquires shared_handle with _cache_version = 1 (still ok - 2 copies of cache exist, but both are actually being used)
DoFn1.start_bundle() acquires shared_handle with _cache_version = 2 (at this point, _cache_version=1 is garbage collected)
DoFn2.start_bundle acquires shared_handle with _cache_version = 1 (this doesn't make sense, it should just jump to _cache_version=2, but the versions are out of sync)
```

Since there's no guarantee that bundle requests are going to get round-robined, over time there could be drift between the various `DoFn`'s `_cache_version`s, leading to extra remote calls. Over time, it is reasonably likely that the cache will degrade to the point where we're not getting any benefit from it.

That's why I suggested still grounding the current `_cache_version` in the current time, but bucketizing it so that it still resolves to a single discrete value per bucket

jaehyeon-kim replied:
Thanks for the clarification. Would you consider the following change is acceptable?

```python
# The wrapper class is needed for a dictionary, because it does not support weak references.
class WeakRefDict(dict):
    pass

class EnrichOrderFn(beam.DoFn):
    def __init__(self):
        self._max_stale_sec = 60
        self._customers = {}
        self._shared_handle = shared.Shared()

    def setup(self):
        # setup is a good place to initialize transient in-memory resources.
        self._customer_lookup = self._shared_handle.acquire(
            self.load_customers, self.set_tag()
        )
 
    def set_tag(self):
        current_ts = datetime.now().timestamp()
        return current_ts - (current_ts % self._max_stale_sec)

    def load_customers(self):
        self._customers = expensive_remote_call_to_load_customers(tag=self.set_tag())
        return WeakRefDict(self._customers)

    def start_bundle(self):
        # Update the lookup table when the time difference exceeds the threshold.
        if self.set_tag() > self._customers["tag"]:
            self._customer_lookup = self._shared_handle.acquire(
                self.load_customers, self.set_tag()
            )

    def process(self, element):
        attr = self._customer_lookup.get(element["customer_id"], {})
        yield {**element, **attr}
```

jaehyeon-kim replied:
Resolved by a separate commit.

damccorm reviewed (commented):
Thanks! Overall, looks pretty good to me, but I had some suggestions

rszper commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+
+# Cache data using a shared object
+
+A cache is a software component that stores data so that future requests for that data can be served faster. Side inputs, stateful `DoFn` and calling an external service can be used for accessing a cache. The Python SDK provide another option in the shared module, which can be more efficient than side inputs in terms of the amount of memory required and simpler than stateful `DoFn`.
```
```suggestion
A cache is a software component that stores data so that future requests for that data can be served faster. To access a cache, you can use side inputs, stateful `DoFn`, and calling an external service. The Python SDK provides another option in the shared module. This option can be more memory-efficient than side inputs and simpler than a stateful `DoFn`.
```

jaehyeon-kim replied:
Resolved by a separate commit.

rszper commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+
+A cache is a software component that stores data so that future requests for that data can be served faster. Side inputs, stateful `DoFn` and calling an external service can be used for accessing a cache. The Python SDK provide another option in the shared module, which can be more efficient than side inputs in terms of the amount of memory required and simpler than stateful `DoFn`.
+
+The samples on this page show you how to use the Shared class of the [shared module](https://beam.apache.org/releases/pydoc/current/apache_beam.utils.shared.html) to enrich elements in both bounded and unbounded `PCollection`s. Two data sets are used in the samples - order and customer. The order records include customer IDs with which customer attributes are added by mapping the customer records ...
```
```suggestion
The examples on this page demonstrate how to use the `Shared` class of the [`shared module`](https://beam.apache.org/releases/pydoc/current/apache_beam.utils.shared.html) to enrich elements in both bounded and unbounded `PCollection` objects. Two data sets are used in the samples: _order_ and _customer_. The order records include customer IDs that customer attributes are added to by mapping the customer records.
```

jaehyeon-kim replied:
Resolved by a separate commit.

rszper commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+
+## Create a cache on a batch pipeline
+
+In this example, the customer cache is loaded as a dictionary in the `setup` method of the `EnrichOrderFn`, and it is used to add customer attributes to the order records. Note that a Shared object encapsulates a weak reference to a singleton instance of the shared resource. Therefore, we need to create a wrapper class since the Python dictionary does not support weak references.
```
```suggestion
In this example, the customer cache is loaded as a dictionary in the `setup` method of the `EnrichOrderFn`. The cache is used to add customer attributes to the order records. Because the Python dictionary doesn't support weak references and a `Shared` object encapsulates a weak reference to a singleton instance of the shared resource, create a wrapper class.
```

jaehyeon-kim replied:
Resolved by a separate commit.

rszper commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+        yield o
+
+
+# wrapper class needed for a dictionary since it does not support weak references.
```
```suggestion
# The wrapper class is needed for a dictionary, because it does not support weak references.
```

jaehyeon-kim replied:
Resolved by a separate commit.

rszper commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+
+## Create a cache and update it regularly on a streaming pipeline
+
+As the batch pipeline example, the customer cache is initialized as a dictionary in the `setup` method. To update the cache, it is refreshed when its timestamp is earlier than the defined threshold value of 5 seconds. The refreshment is implemented in the `start_bundle` method. Once the cached is updated, the order records are enriched with updated customer attributes. Note that it is necessary t ...
```
```suggestion
The customer cache is initialized as a dictionary in the `setup` method. To update the cache, it is refreshed when its timestamp is earlier than the defined threshold value of 5 seconds. The refresh is implemented in the `start_bundle` method. After the cached is updated, the order records are enriched with updated customer attributes. To reload the shared object when it gets refreshed, you must change the `tag` argument of the `acquire` method. This example uses the current timestamp value (`current_ts`).
```

jaehyeon-kim replied:
Resolved by a separate commit.

rszper commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+        yield o
+
+
+# wrapper class needed for a dictionary since it does not support weak references.
```
```suggestion
# The wrapper class is needed for a dictionary, because it doesn't support weak references.
```

jaehyeon-kim replied:
Resolved by a separate commit.

rszper commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,187 @@
+        return WeakRefDict(self._customers)
+
+    def start_bundle(self):
+        # update the lookup table when the time difference exceeds the threshold.
```
```suggestion
        # Update the lookup table when the time difference exceeds the threshold.
```

jaehyeon-kim replied:
Resolved by a separate commit.

jaehyeon-kim commented:
@damccorm I updated the snippets. For the streaming example, it is updated so that the shared object is refreshed when the current tag value exceeds the existing tag value. For this, a tag value is added to the the shared object instead of timestamp.

@rszper I added some more lines in the first paragraph and the description of the streaming pipeline is updated. Please have a look.

damccorm reviewed (approved):
Thanks! This looks good to me - we can merge unless @rszper has further comments

liferoad reviewed (approved):
Great work! Thanks!

rszper commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,93 @@
+
+# Cache data using a shared object
+
+A cache is a software component that stores data so that future requests for that data can be served faster. To access a cache, you can use side inputs, stateful `DoFn`, and calling an external service. The Python SDK provides another option in the shared module. This option can be more memory-efficient than side inputs, simpler than a stateful `DoFn`, and more performant than calling an external ...
```
```suggestion
A cache is a software component that stores data so that future requests for that data can be served faster. To access a cache, you can use side inputs, stateful `DoFn`, and calls to an external service. The Python SDK provides another option in the shared module. This option can be more memory-efficient than side inputs, simpler than a stateful `DoFn`, and more performant than calling an external service, because it does not have to access an external service for every element or bundle of elements. For more details about strategies for caching data using Beam SDK, see the session [Strategies for caching data in Dataflow using Beam SDK](https://2022.beamsummit.org/sessions/strategies-for-caching-data-in-dataflow-using-beam-sdk/) from the 2022 Beam Summit.
```

rszper commented on website/www/site/content/en/documentation/patterns/shared-class.md:
```diff
@@ -0,0 +1,93 @@
+
+## Create a cache and update it regularly on a streaming pipeline
+
+The customer cache is assumed to change over time, and it should be refreshed periodically. Note that, to reload the shared object, you must change the `tag` argument of the `acquire` method. In this example, the refresh is implemented in the `start_bundle` method where it compares the current tag value to the value that is associated with the existing shared object. The `set_tag` method returns  ...
```
```suggestion
Because the customer cache is assumed to change over time, you need to refresh it periodically. To reload the shared object, change the `tag` argument of the `acquire` method. In this example, the refresh is implemented in the `start_bundle` method, where it compares the current tag value to the value that is associated with the existing shared object. The `set_tag` method returns a tag value that is the same within the maximum seconds of staleness. Therefore, if a tag value is greater than the existing tag value, it triggers a refresh of the customer cache.
```

rszper reviewed (approved):
Just some minor wording tweaks; otherwise, LGTM.
